contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,471 @@
+  * `Deallocate`: Called when deleting a container can be used for cleanup
+
+The device plugin is also expected to periodically call the `Heartbeat` function
+exposed by Kubelet and issue a `Registration` request when it either can't reach
+Kubelet or Kubelet answers with a `KO` response.
+
+![Process](./device-plugin.png)
+
+
+### Protobuf specification
+
+```go
+service PluginRegistration {
+	rpc Register(RegisterRequest) returns (RegisterResponse) {}
+	rpc Heartbeat(HeartbeatRequest) returns (HeartbeatResponse) {}
+}
+
+service DevicePlugin {
+	rpc Discover(Empty) returns (stream Device) {}
+	rpc Monitor(Empty) returns (stream DeviceHealth) {}
+
+	rpc Allocate(AllocateRequest) returns (AllocateResponse) {}
+	rpc Deallocate(DeallocateRequest) returns (Empty) {}
+}
+
+message RegisterRequest {
+	// Version of the API the Device Plugin was built against
+	string version = 1;
+	// Name of the unix socket the device plugin is listening on
+	string unixsocket = 2;
+	// Name of the devices the device plugin wants to register
+	// A device plugin can only register one kind of devices
+	string vendor = 3;
+}
+
+message RegisterResponse {
+	// Minimum version the Kubelet API supports.
+	string version = 1;
+	// Kubelet fills this field if it encounters any errors
+	// during the registration process or discover process
+	Error error = 2;
+}
+
+message HeartbeatRequest {
+	string vendor = 1;
+}
+
+message HeartbeatResponse {
+	// Kubelet answers with a string telling the device
+	// plugin to either re-register itself or not
+	string response = 1;
+	// Kubelet fills this field if it encountered any errors
+	Error error = 2;
+}
+
+message AllocateRequest {
+	repeated Device devices = 1;
+}
+
+message AllocateResponse {
+	// List of environment variable to set in the container.
+	repeated KeyValue envs = 1;
+	// Mounts for the container.
+	repeated Mount mounts = 2;
+}
+
+message DeallocateRequest {
+	repeated Device devices = 1;
+}
+
+message Error {
+	bool error = 1;
+	string reason = 2;
+}
+
+// E.g:
+// struct Device {
+//    Kind: "NVIDIA-gpu"
+//    Name: "GPU-fef8089b-4820-abfc-e83e-94318197576e"
+//    Properties: {
+//        "Family": "Pascal",
+//        "Memory": "4G",
+//        "ECC"   : "True",
+//    }
+//}
+//
+message Device {
+	string Kind = 1;
+	string Name = 2;
+	string Health = 3;
+	string Vendor = 4;
+	map<string, string> properties = 5; // Could be [1, 1.2, 1G]
+}
+
+message DeviceHealth {
+	string Name = 1;
+	string Kind = 2;
+	string Vendor = 4;
+	string Health = 3;
+}
+```
+
+## Installation
+
+The installation process should be straightforward to the user, transparent
+and similar to other regular Kubernetes actions.
+The device plugin should also run in containers so that Kubernetes can
+deploy them and restart the plugins when they fail.
+However, we should not prevent the user from deploying a bare metal device
+plugin.
+
+Deploying the device plugins through DemonSets makes sense as the cluster
```

Comment:
i do think this introduces complexity around node drains.  its worth discussing upgrade process.